src/main/flight/autopilot.c

```diff
@@ -91,14 +158,223 @@ void altitudeControl(float targetAltitudeCm, float taskIntervalS, float vertical
+        resetPositionControlParamsNS();
+        resetPositionControlParamsEW();
+        autopilotAngle[AI_ROLL] =  0.0f;
+        autopilotAngle[AI_PITCH] = 0.0f; 
+    } else {
+        // control position
+        // first get xy distances from current location (gpsSol.llh) to target location
+        float nsDistance; // cm, steps of 11.1cm, North of target is positive
+        float ewDistance; // cm, steps of 11.1cm, East of target is positive
+        GPS_distances(&gpsSol.llh, &currentTargetLocation, &nsDistance, &ewDistance);
+        float distanceCm = sqrtf(sq(nsDistance) + sq(ewDistance));
+
+        posHold.pt1Gain = pt1FilterGain(posHold.lpfCutoff, posHold.gpsDataIntervalS);
+
+        // ** Sanity check **
+        // larger threshold if faster at start
+        if (posHold.isStartingNS || posHold.isStartingEW) {
+            posHold.sanityCheckDistance = gpsSol.groundSpeed > 1000 ? gpsSol.groundSpeed : 1000.0f;
+        }
+        // primarily to detect flyaway from no Mag or badly oriented Mag
+        // but must accept some overshoot at the start, especially if entering at high speed
+        if (distanceCm > posHold.sanityCheckDistance) {
+            return false; // must stay within 10m or probably flying away
+            // value at this point is a 'best guess' to detect IMU failure in the event the user has no Mag
+            // if entering poshold from a stable hover, we would only exceed this if IMU was disoriented
+            // if entering poshold at speed, it may overshoot this value and falsely fail, if so need something more complex
+        }
+
+        // ** PIDs **
+        // separate PID controllers for latitude (NorthSouth or ns) and longitude (EastWest or ew)
+
+        // ** P **
+
+        const float nsP = nsDistance * positionPidCoeffs.Kp;
+        const float ewP = ewDistance * positionPidCoeffs.Kp;
+
+        // ** I **
+
+        const float leak = 1.0f - 0.4f * posHold.gpsDataIntervalS; // gpsDataIntervalS is not more than 1.0s
+        if (!posHold.isStartingNS){
+            // only accumulate iTerm after completing the start phase
+            // perhaps need a timeout on the start phase ?
+            posHold.integralNS += nsDistance * posHold.gpsDataIntervalS;
+        } else {
+            // while moving sticks, slowly leak iTerm away, approx 2s time constant
+            posHold.integralNS *= leak;
+        }
+        if (!posHold.isStartingEW){
+            posHold.integralEW += ewDistance * posHold.gpsDataIntervalS;
+        } else {
+            posHold.integralEW *= leak;
+        }
+
+        const float nsI = posHold.integralNS * positionPidCoeffs.Ki;
+        const float ewI = posHold.integralEW * positionPidCoeffs.Ki;
+
+        // ** D ** //
+        // get change in distance in NS and EW directions from gps.c using the `GPS_distances` function
+        // this gives cleaner velocity data than the module supplied GPS Speed and Heading information
```

Comment:
This is strange - GPS should also use  carrier frequency Doppler to estimate speed, so GPS estimate shall be better than simple position difference,

Reply:
I don't know exactly why, but the control accuracy increased greatly when I didn't use the gps-reported speed and the bearing to home.  GPS Speed was particularly erratic and slow to respond.  I think they do some manipulations when it moves slowly.
Things improved by using the bearing to home and the calculated distance to home.  
But position control accuracy is definitely best with the current approach.